modules/upgrade/upgrade.go

```diff
@@ -8,429 +8,288 @@ import (
-			// we already dealt with upgraded
+	var names []string
+	for _, service := range files {
+		name := service.Name()
+		if service.IsDir() || !strings.HasSuffix(name, ".yaml") {
 			continue
 		}
-		log.Info().Str("service", service).Msg("stop service")
-		if err := u.zinit.Stop(service); err != nil {
-			return err
-		}
-	}
-	for _, service := range services {
-		if service == "upgraded" {
-			// we already dealt with upgraded
+
+		name = strings.TrimSuffix(name, ".yaml")
+		// skip self
+		if isIn(name, protected) {
 			continue
 		}
-		if err := u.waitServiceStop(time.Millisecond*500, service); err != nil {
-			return errors.Wrap(err, "failed to stop service")
-		}
+
+		names = append(names, name)
 	}
-	log.Info().Msg("all services stopped")
 
-	if err := executeHook(filepath.Join(flistRoot, string(hookMigrate))); err != nil {
-		log.Error().Err(err).Msg("fail to execute migrate script")
-		return err
+	stopped, err := u.stopMultiple(10*time.Second, names...)
+	if err != nil {
+		return errors.Wrapf(err, "failed to stop services: %+v", names)
 	}
 
-	for _, service := range services {
-		log.Info().Str("service", service).Msg("restart service")
-		if err := u.zinit.Start(service); err != nil {
-			return err
+	// we do a forget so any changes of the zinit config
+	// themselves get reflected once monitored again
+	for _, stopped := range stopped {
+		if err := u.Zinit.Forget(stopped); err != nil {
+			log.Error().Err(err).Str("service", stopped).Msg("error on zinit forget")
 		}
 	}
 
-	if err := executeHook(filepath.Join(flistRoot, string(hookPostStart))); err != nil {
-		log.Error().Err(err).Msg("fail to execute post-start script")
+	if err := copyRecursive(flistRoot, "/", "/bin/upgraded"); err != nil {
 		return err
 	}
 
-	log.Info().Str("flist", upgrade.Flist).Msg(("upgrade applied"))
-
-	return nil
-}
-
-func executeHook(path string) error {
-	name := filepath.Base(path)
-	log.Info().Msgf("prepare to execute %s hook", name)
+	// start all services in the flist
+	for _, name := range names {
+		if err := u.Zinit.Monitor(name); err != nil {
+			log.Error().Err(err).Str("service", name).Msg("error on zinit monitor")
+		}
 
-	info, err := os.Stat(path)
-	if err != nil {
-		if os.IsNotExist(err) {
-			log.Info().Str("script", path).Msgf("%s upgrade hook script not found, skipping", name)
-			return nil
+		// while we totally do not need to call start after monitor but
+		// monitor won't take an action on a monitored service if it's
+		// stopped (but not forgoten). So we call start just to be sure
+		if err := u.Zinit.Start(name); err != nil {
+			log.Error().Err(err).Str("service", name).Msg("error on zinit start")
 		}
-		return err
 	}
 
-	if !isExecutable(info.Mode().Perm()) {
-		return fmt.Errorf("%s exists but is not executable", name)
+	if err := u.BootInfo.Commit(version); err != nil {
+		return err
 	}
 
-	cmd := exec.Command(path)
-	err = cmd.Run()
+	// one last thing. the self upgrade.
+	// 1- Find where upgraded is running from
```

Comment:
copyFile function already allow you to copy a binary file in place even if a process of this binary is already running.
This `exec mv` trick is not necessary and more complex so I would avoid that.

Reply:
I thought that u couldn't solve that, since i found a lot of code regarding updating upgrade itself.

Reply:
Hum could be for the current process you cannot. Let me have another look at the old code see what is what.

Reply:
Well, i remember you had an issue that u couldn't copy over `upgraded` with "file is busy" error or something, hence u had all the logic for the self update.

Reply:
I just check the old code. copyFile indeed already handle the problem of busy file, so no point doing your trick here.

But the logic around upgraded was to allow to restart the daemon itself when there were a new version of it. So the logic was:
- copy files from flist
- if upgraded has a new version => write some flag file => exit(0)
- upgraded then is restarted. I know runs the new version
- if flag file is there, remove it and avoid stopping itself to continue upgrade of the rest.

I think we need to keep this kind of logic. Where the first things we do is to upgrade `upgraded` daemon. Reason is if we need to introduce some change in the upgrade flow. We want upgraded to get this upgrade first before trying to touch any other file.